### DISCUSSION TITLE
remote_server and GUI (eg x11) dependencies

### DISCUSSION THREAD

**Discussion by shish** [zed-industries/zed#19137] [Feature Requests]
ED> It looks like this is only actually an issue at build time and in debug builds, and some kind of magic is already stripping dependencies in release; feel free to ignore, post saved below just in case (that said, I'd love a pointer on how the magic works :eyes: )

--

Right now `remote_server` depends on `gpui` via several different dependency chains, which in turn depends on some GUI libraries that I wouldn't expect to find installed on your average headless linux server or docker container. (Specifically I can't currently remote into my server because it doesn't have libxkbcommon or libxkbcommon-x11 installed (It's also missing libgit2, but that feels like something to deal with separately))

Firstly - are we agreed that `remote_server` should try to keep system-library dependencies minimal, and specifically try to avoid depending on GUI libraries? If the answer is "no we don't care about dependencies, just install X11 stuff on your server and get on with life" then I'll be grumpy but I'll live with it :P (To be clear I don't mind rust dependencies and a huge binary, it's just a pain to install x11 packages on a server where I might not have root access)

If y'all're happy with the general idea of minimising `remote_server`'s dependencies, then I'm happy to give it a shot, but I'm not sure what direction is best - I have a couple of ideas:

- give `gpui` feature flags for the GUI backends - the main `zed` binary depends on `gpui[metal,x11,wayland,windows]`, while `remote_server` can depend on `gpui[headless]` (in practice this'd probably just be two feature flags for x11/wayland, since metal is already implied by macos and windows is already implied by windows)
- looking at the output of `cargo tree --package remote_server --edges=normal | grep zed/crates`, it seems like there are a _lot_ of places where core infrastructure / framework crates (language, lsp, fuzzy, git, ...) depend on `gpui` -- but doing some spot-checks, they all seem to want things like threads and global variables, not windows and widgets. Would it be worth having a separate `app` crate for the application lifecycle / globals / AppContext / etc, and save gpui for the UI?
- Would it be worth having "non-GUI binaries (remote_server, collab, cli... any others?) should not depend on X11 libraries" as part of CI tests, so that things like https://github.com/zed-industries/zed/commit/ba5c1322ced2edf3409c8cb2036d1756c55795d8 don't sneak in by accident in the future?

Do any of those sound like good ideas / the kinds of things somebody relatively-new to the codebase could do?

**Comment by SomeoneToIgnore** [comment]
Those are good ideas, and some of it had landed very recently: https://github.com/zed-industries/zed/pull/19136

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/support
